glean-core/src/metrics/jwe.rs

```diff
@@ -0,0 +1,374 @@
+
+use crate::error_recording::{record_error, ErrorType};
+use crate::metrics::{Metric, MetricType};
+use crate::storage::StorageManager;
+use crate::CommonMetricData;
+use crate::Glean;
+
+const DEFAULT_MAX_CHARS_PER_ELEMENT: usize = 1024;
+
+/// Verifies if a string is [`BASE64URL`](https://tools.ietf.org/html/rfc4648#section-5) compliant.
+///
+/// As such, the string must match the regex: `[a-zA-Z0-9\-\_]*`.
+///
+/// > **Note** As described in the [JWS specification](https://tools.ietf.org/html/rfc7516#section-2),
+/// > the BASE64URL encoding used by JWE discards any padding,
+/// > that is why we can ignore that for this validation.
+///
+/// The regex crate isn't used here because it adds to the binary size,
+/// and the Glean SDK doesn't use regular expressions anywhere else.
+fn validate_base64url_encoding(value: &str) -> bool {
+    let mut iter = value.chars();
+
+    loop {
+        match iter.next() {
+            // We are done, so the whole expression is valid.
+            None => return true,
+            // Valid characters.
+            Some('_') | Some('-') | Some('a'..='z') | Some('A'..='Z') | Some('0'..='9') => (),
+            // An invalid character.
+            Some(_) => return false,
+        }
+    }
+}
+
+/// Representation of a [JWE](https://tools.ietf.org/html/rfc7516).
+#[derive(Serialize)]
+struct Jwe {
+    header: String,
+    key: String,
+    init_vector: String,
+    cipher_text: String,
+    auth_tag: String,
+}
+
+impl Jwe {
+    /// Create a new JWE and validate all elements.
+    ///
+    /// Validation includes checking if each element is valid BASE64URL according the the JWE specification
+    /// and also checking if each element does not exceed DEFAULT_MAX_CHARS_PER_ELEMENT.
+    ///
+    /// **Note** The character limit is our own constraint, not part of the spec.
+    ///
+    /// ## Arguments
+    ///
+    /// * `header` - the JWE Protected Header element.
+    /// * `key` - the JWE Encrypted Key element. May be empy.
+    /// * `init_vector` - the JWE Initialization Vector element.  May be empy.
+    /// * `cipher_text` - the JWE Ciphertext element.
+    /// * `auth_tag` - the JWE Authentication Tag element.  May be empy.
+    fn new<S: Into<String>>(
+        header: S,
+        key: S,
+        init_vector: S,
+        cipher_text: S,
+        auth_tag: S,
+    ) -> Result<Self, (ErrorType, &'static str)> {
+        let header = header.into();
+        let key = key.into();
+        let init_vector = init_vector.into();
+        let cipher_text = cipher_text.into();
+        let auth_tag = auth_tag.into();
+
+        for element in [&header, &cipher_text].iter() {
+            if element.is_empty() {
+                return Err((
+                    ErrorType::InvalidValue,
+                    "Elements `header` and `cipher_text` must not be empty.",
+                ));
+            }
+        }
+
+        for element in [&header, &key, &init_vector, &cipher_text, &auth_tag].iter() {
```

Comment:
Note that only the variable-sized  fields should have at most 1024 characters, as [per our spec](https://docs.google.com/document/d/1nntNIiE6braTGzoKf-lx21OVDd8ssyeIeJu3jnQQfEE/edit?ts=5ee76aa1#heading=h.l4ya7hop8p4q). The others have other sizes (see the SDK API section)